### PR TITLE
chore(cli): Replace `structopt` with `clap` 3

### PR DESCRIPTION
**Pull request by leebenson**
Replaces [structopt](https://github.com/TeXitoi/structopt) with [clap v3](https://github.com/clap-rs/clap), now that structopt is in [maintenance](https://docs.rs/structopt/latest/structopt/#maintenance).

Motivated by unexpected UX in #11442 where root CLI flags + subcommands collide. I haven't tested whether that issue is fixed through this bump, but in any case we should get more knobs to fiddle with.

Signed-off-by: Lee Benson <lee@leebenson.com>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -199,10 +199,10 @@ number_prefix = { version = "0.4.0", default-features = false, features = ["std"
 tui = { version = "0.17.0", optional = true, default-features = false, features = ["crossterm"] }
 
 # Datadog Pipelines
-datadog-search-syntax = { path = "lib/datadog/search-syntax" }
 datadog-filter = { path = "lib/datadog/filter" }
-sha2 = { version = "0.10.2", optional = true }
+datadog-search-syntax = { path = "lib/datadog/search-syntax" }
```
Comment (leebenson) [inline Cargo.toml:203]: Some of these Cargo.toml changes are unrelated. I noticed we were _mostly_ alphabetising deps, and got liberal with the 'Sort Lines' feature in CLion.

File: Cargo.toml [hunk 2]
```diff
@@ -223,9 +224,11 @@ bytes = { version = "1.1.0", default-features = false, features = ["serde"] }
 bytesize = { version = "1.1.0", default-features = false }
 chrono = { version = "0.4.19", default-features = false, features = ["serde"] }
 cidr-utils = { version = "0.5.5", default-features = false }
-csv = { version = "1.1", optional = true }
+clap = { version = "3.1.0", features = ["derive", "env"] }
```
Comment (leebenson) [inline Cargo.toml:227]: This is the only real change in the respective `Cargo.toml`s.

File: lib/vrl/cli/src/cmd.rs [hunk 3]
```diff
@@ -6,37 +6,37 @@ use std::{
...
 
     /// Print the (modified) event object instead of the result of the final expression. Setting
     /// this flag is equivalent to using `.` as the final expression.
-    #[structopt(short = "o", long)]
+    #[clap(short = 'o', long)]
     print_object: bool,
 
     /// The timezone used to parse dates.
-    #[structopt(short = "tz", long)]
+    #[clap(short = 't', long)]
```
Comment (leebenson) [inline lib/vrl/cli/src/cmd.rs:39]: **⚠️ Breaking change.**

In clap 3, short is a `char` type.

I chose `t` since it's the first character, but it may be that `z` is more commonly used for timezones.

Strong opinions welcome.

File: src/cli.rs [hunk 4]
```diff
@@ -10,25 +10,23 @@ use crate::tap;
...
         let version = get_version();
-        let app = Opts::clap().version(version.as_str()).global_settings(&[
-            AppSettings::ColoredHelp,
-            AppSettings::InferSubcommands,
-            AppSettings::DeriveDisplayOrder,
-        ]);
-        Opts::from_clap(&app.get_matches())
+        let app = Opts::command()
+            .version(version.as_str())
+            .global_setting(AppSettings::DeriveDisplayOrder);
```
Comment (leebenson) [inline src/cli.rs:28]: `AppSettings::ColoredHelp` is a default in clap 3, so this was removed.

 `AppSettings::InferSubcommands` has been deprecated in favour of https://docs.rs/clap/latest/clap/struct.App.html#method.infer_subcommands

It briefly crossed my mind that this may be a culprit for the collision between the `-c` flag and `config` subcommand in https://github.com/vectordotdev/vector/pull/11442, although they should be distinct domains, so may be nothing.

In any case, I don't know how useful (or encouraged) it is to type, say, `vector vr` instead of `vector vrl`. I'd prefer explicit names personally given the growing number of subcommands.

Again, strong opinions welcome.

File: src/cli.rs [hunk 5]
```diff
@@ -10,25 +10,23 @@ use crate::tap;
...
+        Opts::from_arg_matches(&app.get_matches()).expect("couldn't initialize CLI. Please report.")
```
Comment (leebenson) [inline src/cli.rs:29]: `Opts::from_clap` is deprecated in favour of `Opts::from_arg_matches`, which now returns `Result<Self, Error>`.

I'm not 100% on which conditions would cause this to panic, other than misconfigured field annotations, so I've deemed this an 'us' problem and elected to panic with a message to report.

File: src/cli.rs [hunk 6]
```diff
@@ -59,83 +57,83 @@ impl Opts {
...
     /// If zero files are specified the default config path
     /// `/etc/vector/vector.toml` will be targeted.
-    #[structopt(
+    #[clap(
         name = "config",
         short,
         long,
         env = "VECTOR_CONFIG",
-        use_delimiter(true)
+        use_value_delimiter(true)
```
Comment (leebenson) [inline src/cli.rs:72]: `use_delimiter` is deprecated in clap 3.

This is chiefly what motivated this PR, per #11442

File: lib/vrl/cli/src/cmd.rs (hunk 3 above)
Comment (jszwedko) [reply lib/vrl/cli/src/cmd.rs:39]: I like `-z` just because it would free up `-t`, a more common letter, for something else.

File: src/cli.rs (hunk 4 above)
Comment (jszwedko) [reply src/cli.rs:28]: Yeah, agreed, I like the full command names.

File: src/cli.rs (hunk 5 above)
Comment (jszwedko) [reply src/cli.rs:29]: Yeah, that method doesn't document itself very well. Looking at `clap::Error` the kinds are these:

https://docs.rs/clap/latest/clap/error/enum.ErrorKind.html

Which makes me think that maybe this would return an error simply if the user passed, for example, an invalid subcommand? In which case, we should propagate it up?

If we do choose to panic here, it'd be nice to include the `Err` value itself to have more of a clue of what failed.

**Review by jszwedko — COMMENTED** [review]
Looks good! Thanks for the explanatory review comments. I left a few comments below, one blocking, but overall 👍 to this change.

**Comment by spencergilbert** [comment]
Oh and obvious upgrade guide needed for the breaking changes 😄

File: src/cli.rs (hunk 5 above)
Comment (leebenson) [reply src/cli.rs]: Ah, thanks for looking -- yeah, propagating makes sense.

Adjusted in https://github.com/vectordotdev/vector/pull/11470/commits/c6a1da9e9a20c93db3de176c1bd31f63aa2454b4 to print a [friendlier error](https://docs.rs/clap/latest/src/clap/error/kind.rs.html#401-437) and return `exitcode::USAGE`

**Review by jszwedko — COMMENTED** [review]
Looks good! I agree with Spencer (good note!) that we should document the breaking changes in the upgrade guide for 0.21.0.

**Review by JeanMertz — APPROVED** [review]
Sweet! This LGTM
